core/src/main/java/io/grpc/ClientCallImpl.java

```diff
@@ -0,0 +1,349 @@
+    checkState(state == State.INIT, "Already started");
+    this.userAgent = userAgent;
+    return this;
+  }
+
+  @Override
+  public synchronized void start(ClientCall.Listener<RespT> callListener, Headers headers) {
+    checkState(state == State.INIT, "Already started");
+    state = State.STARTED;
+
+    // We need to keep a copy of the call listener because the stream may not be set up in time.
+    // This would happen if the transport is not ready, but the call deadline is reached, and we
+    // need to notify the callListener.
+    this.callListener = checkNotNull(callListener, "No call listener provided");
+
+    ClientStreamListener streamListener = new ClientStreamListenerImpl();
+    Status errorStatus = prepareHeaders(headers, callOptions, userAgent);
+    if (errorStatus != null) {
+      closeCallPrematurely(streamListener, errorStatus);
+      return;
+    }
+
+    boolean started = startInternal(streamListener, headers);
+
+    // Start the deadline timer after stream creation because it will close the stream
+    if (started && headers.containsKey(TIMEOUT_KEY)) {
+      deadlineCancellationFuture = startDeadlineTimer(headers.get(TIMEOUT_KEY));
+    }
+  }
+
+  /**
+   * Modifies headers to send for this call.
+   *
+   * @return null on success, or a failure status
+   */
+  private static Status prepareHeaders(
+      Headers headers, CallOptions callOptions, @Nullable String userAgent) {
+    // Fill out timeout on the headers
+    headers.removeAll(TIMEOUT_KEY);
+    // Convert the deadline to timeout. Timeout is more favorable than deadline on the wire
+    // because timeout tolerates the clock difference between machines.
+    Long deadlineNanoTime = callOptions.getDeadlineNanoTime();
+    long timeoutMicros = 0;
+    if (deadlineNanoTime != null) {
+      timeoutMicros = TimeUnit.NANOSECONDS.toMicros(deadlineNanoTime - System.nanoTime());
+      if (timeoutMicros <= 0) {
+        return Status.DEADLINE_EXCEEDED;
+      }
+      headers.put(TIMEOUT_KEY, timeoutMicros);
+    }
+
+    // Fill out the User-Agent header.
+    headers.removeAll(HttpUtil.USER_AGENT_KEY);
+    if (userAgent != null) {
+      headers.put(HttpUtil.USER_AGENT_KEY, userAgent);
+    }
+    return null;
+  }
+
+  private boolean startInternal(final ClientStreamListener streamListener, final Headers headers) {
+    ClientTransport transport;
+    try {
+      transport = transportProvider.get();
+    } catch (RuntimeException ex) {
+      // TODO(carl-mastrangelo): Eventually remove this catch statement once obtainActiveTransport
+      // no longer throws.
+      closeCallPrematurely(streamListener, Status.fromThrowable(ex));
+      return false;
+    }
+    if (transport == null) {
+      closeCallPrematurely(
+          streamListener, Status.UNAVAILABLE.withDescription("Channel is shutdown"));
+      return false;
+    }
+
+    try {
+      synchronized (this) {
```

Comment:
This was necessary in the retrying case, since start was called possibly from another thread.  Considering that this isn't incorrect, and that this already holds the lock, I would prefer to leave it in place, so I don't forget it in the followup PR.